pkg/leader/leaderelector.go

```diff
@@ -0,0 +1,97 @@
+package leader
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+const (
+	configMapName = "assisted-service-leader-election-helper"
+)
+
+type ElectorInterface interface {
+	StartLeaderElection(ctx context.Context, leaderFunc func(), stopLeaderFunc func()) error
+}
+
+type FakeElector struct{}
+
+func (f *FakeElector) StartLeaderElection(ctx context.Context, leaderFunc func(), stopLeaderFunc func()) error {
+	go leaderFunc()
+	return nil
+}
+
+var _ ElectorInterface = &Elector{}
+
+type Elector struct {
+	log       *logrus.Logger
+	namespace string
+	kube      *kubernetes.Clientset
+}
+
+func NewElector(kubeClient *kubernetes.Clientset, namespace string, logger *logrus.Logger) *Elector {
+	return &Elector{log: logger, namespace: namespace, kube: kubeClient}
+}
+
+func (l *Elector) StartLeaderElection(ctx context.Context, leaderFunc func(), stopLeaderFunc func()) error {
+
+	resourceLock, err := l.createResourceLock(configMapName)
+	if err != nil {
+		return err
+	}
+
+	leaderElector, err := l.createLeaderElector(resourceLock, leaderFunc, stopLeaderFunc)
+	if err != nil {
+		return err
+	}
+
+	l.log.Info("Attempting to acquire leader lease")
+	go leaderElector.Run(ctx)
+
+	return nil
+}
+
+func (l *Elector) createResourceLock(name string) (resourcelock.Interface, error) {
+	// Leader id, needs to be unique
+	id, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	id = id + "_" + string(uuid.NewUUID())
+
+	return resourcelock.New(resourcelock.ConfigMapsResourceLock,
+		l.namespace,
+		name,
+		l.kube.CoreV1(),
+		l.kube.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity: id,
+		})
+}
+
+func (l *Elector) createLeaderElector(resourceLock resourcelock.Interface, leaderFunc func(), stopLeaderFunc func()) (*leaderelection.LeaderElector, error) {
+	return leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+		Lock:          resourceLock,
+		LeaseDuration: 15 * time.Second,
```

Comment:
it should be parameters from env

Reply:
thought to start it this way and then to add config as a next step

Reply:
don't see any reason to change default values. If it can be done in another pr, will be better

Reply:
why delay something so simple?

Reply:
cause i testing it will not be so simple :) i prefer to have it another day. Till now it is blocker, less things to test better for me.

Reply:
will try to do it. let's see